[Fanny Jian] iP

fannyjian opened the pull request:
# l u n a 🌙

> “The Moon is the first milestone on the road to the stars” – Arthur C. Clarke

Luna is your trusty personal assistant bot that's sure to help you complete all your tasks! It's 
* easy to use 🍃
* ~~super extra~~ aesthetically pleasing! 🌷


Get started with Luna today! You just need to:
1. Download **luna.jar** from [Luna v0.1](https://github.com/fannyjian/ip/releases)  
2. Double-click on the file
3. Follow Luna commands to interact and add your tasks ☀️

## features

- [ ] Managing tasks ⛅️
- [x] Managing deadlines (coming soon) 🌨
- [ ] Reminders (coming soon) 🌩

---
Sneak peak at Luna's `main` method!
```ruby
 public static void main(String[] args) {
     new Luna("data/luna.txt").run();
 }
```

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

fannyjian committed:
Add Greet, Echo, Exit, Add and List

fannyjian committed:
Add mark and unmark

fannyjian committed:
Refactor

fannyjian committed:
Add ToDo, Event and Deadline

fannyjian committed:
Add Automated Text UI Testing

fannyjian committed:
Add Handle Errors

fannyjian committed:
Add Delete

fannyjian committed:
Update error message

fannyjian committed:
Update bot behavior

fannyjian committed:
Fix logo for testing

fannyjian committed:
Refactor Duke to Luna

fannyjian committed:
Add save function

fannyjian committed:
Add command list

fannyjian committed:
Add Dates and Times formatting

fannyjian committed:
Resolve merge conflict between master and branch-Level-8

fannyjian committed:
Refactor code to use more OOP

fannyjian committed:
Divide classes into relevant packages

fannyjian committed:
Merge add-gradle-support

fannyjian committed:
Add Gradle

fannyjian committed:
Add JUnit tests for Parser and Task classes

fannyjian committed:
Add JavaDoc to all classes

fannyjian committed:
Refactor command package

fannyjian committed:
Created new commands package

fannyjian committed:
Refactor tasks package

fannyjian committed:
Recreated tasks package

fannyjian committed:
Follow Coding Standard

fannyjian committed:
Add Find Function

fannyjian committed:
Resolve merge conflicts with branch-A-CodingStandard

fannyjian committed:
Resolve merge conflict with branch-Level-9

fannyjian committed:
Format all for loops to follow coding standard

fannyjian committed:
Add new Luna exception cases

fannyjian committed:
Refactor exception package

fannyjian committed:
Include new commands and exception cases for testing

fannyjian committed:
Update commands in welcome message

fannyjian committed:
Update Javadocs to follow coding standards

fannyjian committed:
Add CheckStyle and edit code to follow coding standards

fannyjian committed:
Fix more coding standard violations

fannyjian committed:
Set up JavaFX

rjkoh commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+     *
+     * @param des Description of task.
+     */
+    public AddCommand(String des) {
```
Perhaps you can consider a more understandable and intuitive variable name like "description"? instead of shortening it to "des"

rjkoh commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,193 @@
+package parser;
+
+// Import Commands
+import command.AddCommand;
+import command.Command;
+import command.DeleteCommand;
+import command.ExitCommand;
+import command.FindCommand;
+import command.ListCommand;
+import command.UpdateCommand;
+
+// Import Luna Exceptions
+import exception.LunaException;
+import exception.LunaInvalidCommandException;
+import exception.LunaInvalidDateException;
+import exception.LunaInvalidDescriptionException;
+import exception.LunaInvalidIndexException;
+
+// Import other exceptions
+import java.time.format.DateTimeParseException;
+
+// Import Tasks;
+import task.Deadline;
+import task.Event;
+import task.Task;
+import task.TaskList;
+import task.Todo;
+
+// Imports for dates
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
```
I like how you section the imports to make the purpose of those imports clear

rjkoh commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,193 @@
+     * @return Command corresponding to the user's instruction.
+     * @throws LunaException.
+     */
+    public static Command parse(String cmd) throws LunaException {
+        String[] cmdSplit = cmd.split(" ", 2);
+        Command c;
```
Perhaps you can consider using more intuitive variable names like "command", "splitCommand" etc, and not use c (unless for iterator variables)

rjkoh commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,193 @@
+        } else if (txt.startsWith("[D]")) {
+            // Abstract task description and date
+            String[] split = txt.split("] ");
+            String [] desSplit = split[1].split(" BY ");
+            String des = desSplit[0];
+            String by = desSplit[1].substring(0, 11);
+
+            // Create date
+            LocalDate date = LocalDate.parse(by, DateTimeFormatter.ofPattern("dd-MMM-yyyy"));
+
+            // Create new Deadline
+            Task tsk = new Deadline(des, date);
+
+            // Update status
+            if (split[0].substring(3).equals("[✧")) {
+                tsk.setStatusIcon(true);
+            }
+            return tsk;
+        } else if (txt.startsWith("[E]")) {
+            // Abstract task description and date
+            String[] split = txt.split("] ");
+            String [] desSplit = split[1].split(" AT ");
+            String des = desSplit[0];
+            String at = desSplit[1].substring(0, 11);
+
+            // Create date
+            LocalDate date = LocalDate.parse(at, DateTimeFormatter.ofPattern("dd-MMM-yyyy"));
+
+            // Create new Event
+            Task tsk = new Event(des, date);
+
+            // Update status
+            if (split[0].substring(3).equals("[✧")) {
+                tsk.setStatusIcon(true);
+            }
+            return tsk;
+        } else {
+            return null;
+        }
+    }
```
I like your simple javadocs, and the comments to breakdown the process for readers

rjkoh commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,187 @@
+        System.out.println(SEP + "\nHello. ⛅️\n   This is\n" + logo);
+
+        // Print available commands
+        System.out.println("  Luna commands"
+                + "\n    🌸 list                             | View all tasks on your agenda"
+                + "\n    🌷 todo <task>                      | Add a task to your agenda"
+                + "\n    🌺 deadline <task> /by <yyyy-mm-dd> | Add a task to complete by the specified deadline"
+                + "\n    🌹 event <event> /at <yyyy-mm-dd>   | Add an event on the specified date"
+                + "\n    🪷 mark <num>                       | Mark the (num)th item in your list as completed"
+                + "\n    🌻 unmark <num>                     | Mark the (num)th item in your list as uncompleted"
+                + "\n    💐 find <keyword>                   | Find a task by searching for a keyword"
+                + "\n    🥀 bye                              | Quit Luna\n");
```
I like this feature and the improved aesthetics!

rjkoh commented on src/test/java/tasks/TaskTest.java:
```diff
@@ -0,0 +1,27 @@
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TaskTest {
+    @Test
+    public void TaskTestOne(){
+        Task test = new Event("event", "2002-02-02");
+        test.setStatusIcon(true);
+        assertEquals(test.toString(), "[E][✧] event AT 02 Feb 2002");
+    }
+
+    @Test
+    public void TaskTestTwo(){
+        Task test = new Deadline("task", "2015-03-01");
+        test.setStatusIcon(true);
+        assertEquals(test.toString(), "[D][✧] task BY 01 Mar 2015");
+    }
+
+    @Test
+    public void TaskTestThree(){
+        Task test = new Todo("task");
+        test.setStatusIcon(false);
+        assertEquals(test.toString(), "[T][ ] task");
```
perhaps you could consider following the naming convention for tests

rjkoh reviewed (commented):
overall, a job well done! I like the splitting of commands into their own classes and the different exceptions, along with the added effort into beautifying the UI!

fannyjian committed:
Add GUI

fannyjian committed:
Rebuild scene with FXML

fannyjian committed:
Update GUI

pratham31012002 commented on src/main/java/Luna.java:
```diff
@@ -0,0 +1,62 @@
+        tasks = new TaskList(storage.load(ui));
+    }
+
+//    /**
+//     * Starts the chatbot by reading commands from and executing them.
+//     */
+//    public void run() {
+//        ui.showWelcome();
+//
+//        boolean isExit = false;
+//        while (!isExit) {
+//            try {
+//                String fullCommand = ui.readCommand();
+//                Command c = Parser.parse(fullCommand);
+//                c.execute(tasks, ui, storage);
+//                isExit = c.isExit();
+//            } catch (LunaException e) {
+//                ui.showError(e);
+//            }
+//        }
+//    }
```
Perhaps you can decide whether to keep or remove these comments?

pratham31012002 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,31 @@
+            fxmlLoader.<MainWindow>getController().setLuna(luna);
+            stage.show();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Perhaps the exception can be handled in a different way instead of printing the stack trace?

pratham31012002 commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    @Override
+    public boolean isExit() {
+        return false;
+    }
```
Perhaps this can be added in the parent class and only override in the exit command class to avoid redundancy? And maybe do the same for other commands?

pratham31012002 commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+            tasks.add(this.task);
+            storage.updateStorage(tasks);
+            return ui.showAdded(tasks, this.task);
+        } catch (LunaException e) {
+            return ui.showError(e);
```
In case `storage.updateStorage(tasks)` throws an exception, do you still want the `task` to be added to `tasks`?

pratham31012002 commented on src/main/java/command/UpdateCommand.java:
```diff
@@ -0,0 +1,53 @@
+     * @param type Type of update to be performed.
+     * @param num Index of task in list displayed to user.
+     */
+    public UpdateCommand(String type, int num) {
+        this.type = type;
+        this.num = num - 1;
+    }
```
I like how you have made a single command for both marking and unmarking.

pratham31012002 commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+     * @param des Description of Task.
+     * @param date Date to be completed by.
+     */
+    public AddCommand(String type, String des, LocalDate date) {
```
I like how you have used OOPs principles to distinguish between todo and deadline and event tasks.

pratham31012002 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,192 @@
+    public static Command parse(String cmd) throws LunaException {
+        String[] cmdSplit = cmd.split(" ", 2);
+        Command c;
+        switch (cmdSplit[0]) {
+            case "todo":
```
Perhaps you can change the indentation of the case statement to the same as the switch statement?

pratham31012002 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,192 @@
+     * @return A LocalDate in dd-MMM-yyyy format.
+     * @throws LunaException.
+     */
+    public static LocalDate parseDate(String txt) throws LunaException {
```
I like how you have parsed the entire input in the same file.

pratham31012002 reviewed (commented):
Overall, I think your project is really clean with good use of OOP principles except for a few nits.

fannyjian committed:
Update code to follow coding standards

fannyjian committed:
Edit addLine function

fannyjian committed:
Enhance GUI appearance

Long messages are cut short when new commands are sent by user such that the full message is unable to be seen. The GUI also looks plain.

Let's
* update the DialogBox such that the preferred height and width is used to display messages
* modify the background colors and fonts

Using the preferred height and width at all times ensure that the message will be shown as its original.

fannyjian committed:
Add assertions

fannyjian committed:
Improve GUI appearance and UI welcome message

fannyjian committed:
Improve welcome message

fannyjian committed:
Improve code quality

fannyjian committed:
Add streams

fannyjian committed:
Merge pull request #1 from fannyjian/branch-A-Assertions

Add assertions

fannyjian committed:
Merge branch 'master' into branch-A-CodeQuality

fannyjian committed:
Merge pull request #2 from fannyjian/branch-A-CodeQuality

Improve code quality

fannyjian committed:
Resolve merge conflict with master

fannyjian committed:
Merge pull request #3 from fannyjian/branch-A-Streams

Add streams

fannyjian committed:
Improve GUI and fix bugs

fannyjian committed:
Fix date parser

fannyjian committed:
Fix coding violations

fannyjian committed:
Add more error handling

fannyjian committed:
Implement C-DetectDuplicates

fannyjian committed:
Update exit timer

fannyjian committed:
Add screenshot of GUI

fannyjian committed:
Fix minor bug

fannyjian committed:
Add Luna user guide

fannyjian committed:
Edit title emoji

fannyjian committed:
Edit title emoji

fannyjian committed:
Centralise title

fannyjian committed:
Update README.md

fannyjian committed:
Update jar file name

fannyjian committed:
Add Getting Started section

fannyjian committed:
Add website link

fannyjian committed:
Add screenshot

fannyjian committed:
Centralise all headings

fannyjian committed:
Remove website link

fannyjian committed:
Edit HTML tags

fannyjian committed:
Add collapsible contents list

fannyjian committed:
Edit contents links

fannyjian committed:
Add HTML tags for content links

fannyjian committed:
Improve GUI appearance

fannyjian committed:
Update screenshot of GUI

fannyjian committed:
Update GUI screenshot

fannyjian committed:
Update README.md

fannyjian committed:
Update HTML tags

fannyjian committed:
Update README.md

fannyjian committed:
Fix content to display properly on Github pages

fannyjian committed:
Remove Features section

fannyjian committed:
Update README.md

fannyjian committed:
Create _config.yml

fannyjian committed:
Update content to fit new theme

fannyjian committed:
Update _config.yml

fannyjian committed:
Edit website info

fannyjian committed:
Update README.md

fannyjian committed:
Improve readability

fannyjian committed:
Add website content

fannyjian committed:
Update image link

fannyjian committed:
Edit title

fannyjian committed:
Revert "Edit title"

This reverts commit 2817a5b5d115cb936e5721132c0b726b7742dc32.

fannyjian committed:
Change Ui.png to a cleaner screenshot

fannyjian committed:
Use higher quality image

fannyjian committed:
Update README.md

fannyjian committed:
Update GUI appearance

fannyjian committed:
Add table of contents

fannyjian committed:
Update title and description

fannyjian committed:
Add background color

fannyjian committed:
Update _config.yml

fannyjian committed:
Add title style

fannyjian committed:
Remove unnecessary text

fannyjian committed:
Update README.md

fannyjian committed:
Fix numbering and format issues

fannyjian committed:
Add website link

fannyjian committed:
Improve code quality

fannyjian committed:
Change JavaFX version to 11
